### PR TITLE
Use tapi for printing a TBD file as a YAML file.

### PR DESCRIPTION
**Pull request by huonw**
This imports a big chunk of a newer version of https://github.com/ributzka/tapi as unchanged as possible so that it is as easy to switch over to using that directly at some point in the future. The first commit does this import, and all interesting interactions are in the later ones.

### CODE REVIEW

**Comment by huonw** [comment]
This is still a slightly work in progress, as there's still a few details I want to nail down.

**Review by jrose-apple — COMMENTED** [review]
I'm not fully up to speed on the format or on how you're generating the content, so I can't really provide much in the way of feedback. I do think it's a little weird how much the code is reaching inside `tapi::internal`, though. It seems like if you're meant to use it from outside the tapi/ folder, it shouldn't be internal. But maybe that's a question for @ributzka instead.

(Also, if every symbol kind we use today is a GlobalSymbol, it doesn't really seem worth specifying that at all the call sites right now.)

**Comment by huonw** [comment]
> I do think it's a little weird how much the code is reaching inside tapi::internal, though. It seems like if you're meant to use it from outside the tapi/ folder, it shouldn't be internal. But maybe that's a question for @ributzka instead.

I think this is mostly an artefact of tapi not having any other clients at the moment, but as you say @ributzka would have to answer this. My thinking is that since we're vendoring the code directly at the moment, it's okay to use the internal details, and, by keeping essentially the same code structure, a necessary update (e.g. a bug in the formatting code, or a new format detail to account for) is more feasible than having a complete reimplementation.

> (Also, if every symbol kind we use today is a GlobalSymbol, it doesn't really seem worth specifying that at all the call sites right now.)

Not quite every symbol: there's ObjectiveCClasses too. But yes, having the argument default to GlobalSymbol seems better.

File: lib/TBDGen/TBDGen.cpp [hunk 1]
```diff
@@ -359,19 +390,32 @@ void TBDGenVisitor::visitEnumDecl(EnumDecl *ED) {
...
   UniversalLinkageInfo linkInfo(target, opts.HasMultipleIGMs, isWholeModule);
 
-  TBDGenVisitor visitor(symbols, target, linkInfo, M, opts);
+  tapi::internal::InterfaceFile file;
+  file.setFileType(tapi::internal::FileType::TBD_V3);
+  file.setInstallName(opts.InstallName);
+  // FIXME: proper version
+  file.setCurrentVersion(tapi::internal::PackedVersion(1, 0, 0));
+  // FIXME: pull this out of moduleformat?
+  file.setSwiftABIVersion(5);
```
[the remark this answers is not in the document]
Comment (huonw) [reply lib/TBDGen/TBDGen.cpp]: Ah, I guess we'll hard code it here too, somewhere. :(

File: lib/TBDGen/TBDGen.cpp [hunk 2]
```diff
@@ -403,29 +445,27 @@ static void enumeratePublicSymbolsAndWrite(ModuleDecl *M, FileUnit *singleFile,
...
-    std::vector<StringRef> sorted;
-    for (auto &symbol : symbols)
-      sorted.push_back(symbol.getKey());
-    std::sort(sorted.begin(), sorted.end());
-    for (const auto &symbol : sorted) {
-      *os << symbol << "\n";
-    }
+    tapi::internal::YAMLWriter writer;
+    writer.add(std::unique_ptr<tapi::internal::DocumentHandler>(
+        new tapi::internal::stub::v3::YAMLDocumentHandler));
```
Comment (jrose-apple) [inline lib/TBDGen/TBDGen.cpp]: Nitpick: `llvm::make_unique<tapi::internal::stub::v3::YAMLDocumentHandler>()`. (`unique_ptr` supports upcast conversions.)

**Review by jrose-apple — APPROVED** [review]
Seems like a reasonable change, then, since it's basically doing the same thing as before, just in a more structured way.

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test Linux

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test

**Pull request merged**